core/watcher/watchertest/watcher.go

```diff
@@ -44,3 +47,49 @@ func DirtyKill[T any](c *gc.C, w watcher.Watcher[T]) {
 		c.Logf("ignoring failed to close for watcher")
 	}
 }
+
+// WatcherC embeds a gocheck.C and adds methods to help
+// verify the behaviour of generic watchers.
+type WatcherC[T any] struct {
```

Comment:
I think we should combine `WatcherC` with https://github.com/juju/juju/pull/17720/files#diff-19c9688f769da1df646db29bad1125c327faf336bdfe0af4cb4ba1585afc1916R35-R115